beacon-chain/sync/rpc_blob_sidecars_by_root.go

```diff
@@ -0,0 +1,86 @@
+	denebFork := params.BeaconConfig().DenebForkEpoch
+	var reqWindow primitives.Epoch
+	if current > params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest {
+		reqWindow = current - params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest
+	}
+	if finalized >= reqWindow && finalized > denebFork {
+		return finalized
+	}
+	if reqWindow >= finalized && reqWindow > denebFork {
+		return reqWindow
+	}
+	return denebFork
+}
+
+// blobSidecarByRootRPCHandler handles the /eth2/beacon_chain/req/blob_sidecars_by_root/1/ RPC request.
+// spec: https://github.com/ethereum/consensus-specs/blob/a7e45db9ac2b60a33e144444969ad3ac0aae3d4c/specs/deneb/p2p-interface.md#blobsidecarsbyroot-v1
+func (s *Service) blobSidecarByRootRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
+	ctx, span := trace.StartSpan(ctx, "sync.blobSidecarByRootRPCHandler")
+	defer span.End()
+	ctx, cancel := context.WithTimeout(ctx, ttfbTimeout)
+	defer cancel()
+	SetRPCStreamDeadlines(stream)
+	log := log.WithField("handler", p2p.BlobSidecarsByRootName[1:]) // slice the leading slash off the name var
+	ref, ok := msg.(*types.BlobSidecarsByRootReq)
+	if !ok {
+		return errors.New("message is not type BlobSidecarsByRootReq")
+	}
+	minReqEpoch := blobMinReqEpoch(s.cfg.chain.FinalizedCheckpt().Epoch, slots.ToEpoch(s.cfg.clock.CurrentSlot()))
+	blobIdents := *ref
+	for i := range blobIdents {
+		root, idx := bytesutil.ToBytes32(blobIdents[i].BlockRoot), blobIdents[i].Index
+		scs, err := s.cfg.beaconDB.BlobSidecarsByRoot(ctx, root)
+		if err != nil {
+			if errors.Is(err, db.ErrNotFound) {
+				log.WithError(err).Errorf("error retrieving BlobSidecar, root=%x, index=%d", root, idx)
+				continue
+			}
+			log.WithError(err).Debugf("error retrieving BlobSidecar, root=%x, index=%d", root, idx)
+			s.writeErrorResponseToStream(responseCodeServerError, types.ErrGeneric.Error(), stream)
+			return err
+		}
+		if idx >= uint64(len(scs)) {
+			continue
+		}
+		sc := scs[idx]
+
+		// If any root in the request content references a block earlier than minimum_request_epoch,
+		// peers MAY respond with error code 3: ResourceUnavailable or not include the blob in the response.
+		if slots.ToEpoch(sc.Slot) < minReqEpoch {
+			s.writeErrorResponseToStream(responseCodeResourceUnavailable, types.ErrBlobLTMinRequest.Error(), stream)
+			log.WithError(types.ErrBlobLTMinRequest).
+				Debugf("requested blob for block %#x before minimum_request_epoch", blobIdents[i].BlockRoot)
+			return types.ErrBlobLTMinRequest
+		}
+		SetStreamWriteDeadline(stream, defaultWriteDuration)
+		if chunkErr := WriteBlobSidecarChunk(stream, s.cfg.chain, s.cfg.p2p.Encoding(), sc); chunkErr != nil {
+			log.WithError(chunkErr).Debug("Could not send a chunked response")
+			s.writeErrorResponseToStream(responseCodeServerError, types.ErrGeneric.Error(), stream)
+			tracing.AnnotateError(span, chunkErr)
+			return chunkErr
+		}
+		s.rateLimiter.add(stream, 1)
+	}
+	closeStream(stream, log)
```

Comment:
Shouldn't this be a defer as soon as the stream is opened? Seems like any of the early terminations on error wouldnt close the stream. 

Also, wouldn't the caller be responsible for closing the stream? Perhaps you could add some commentary or something to explain why this is here in the way that it is.

Reply:
If you look at other RPC handlers, they all do the same thing. The reason is that the error handling functions independently call `closeStream`, so calling it in a defer would cause the stream to be closed twice. I want to refactor the dispatch code to take care of this after the handler completes and remove the close from the error handlers, since we should always close the stream in both error and successful paths.

Reply:
I'm going to do some work to fix this for all RPCs in develop. In the meantime the context check adds another place where we need to explicitly call closeStream. I thought this was cleaner than my other idea, which was to defer with an enclosed bool to track whether an error method had been called. Once I get the other PR into develop and we rebase, I'll update this code to defer the stream close. Hopefully that doesn't block this PR.